### ISSUE TITLE
MultiStepLR is broken with .step(epoch=xyz)

### ISSUE THREAD

**Issue by alsrgv** [pytorch/pytorch#33229]
## 🐛 Bug

MultiStepLR is broken with `.step(epoch=xyz)`:

```
import torch
t = torch.tensor(1.0, requires_grad=True)
opt = torch.optim.SGD([t], lr=0.01)
s = torch.optim.lr_scheduler.MultiStepLR(opt, milestones=[19])
print('lr', opt.param_groups[0]['lr'])

opt.step()
s.step(0)
print('lr', opt.param_groups[0]['lr'])
```

Outputs:
```
lr 0.01
lr 0.001
```

I'm aware that `.step(epoch=xyz)` is deprecated, but it's used by third-party libraries, like https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/training_loop.py#L341.

cc @vincentqb @williamFalcon ^

I've debugged the issue and it seems to be caused by `bisect_right()` not playing well with `Counter`: https://github.com/pytorch/pytorch/blob/master/torch/optim/lr_scheduler.py#L399

This can be fixed by casting self.milestones to list before bisection.  If that's an acceptable fix, I'm happy to send a PR.

## Environment

```
Collecting environment information...
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 4.8.5-4ubuntu8) 4.8.5
CMake version: version 3.12.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.1.243
GPU models and configuration:
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti

Nvidia driver version: 440.33.01
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.3

Versions of relevant libraries:
[pip3] numpy==1.17.2
[pip3] pytorch-lightning==0.5.3.2
[pip3] torch==1.4.0
[pip3] torch2trt==0.0.3
[pip3] torchvision==0.6.0a0
[conda] Could not collect
```

**Comment by zhangguanheng66** [comment]
@vincentqb

**Comment by vincentqb** [comment]
> I've debugged the issue and it seems to be caused by `bisect_right()` not playing well with `Counter`: https://github.com/pytorch/pytorch/blob/master/torch/optim/lr_scheduler.py#L399
> 
> This can be fixed by casting self.milestones to list before bisection. If that's an acceptable fix, I'm happy to send a PR.

Converting to list is a solution (#33229), but we could also remove counter #33357 introduced in #26423. Thoughts?

> I'm aware that `.step(epoch=xyz)` is deprecated, but it's used by third-party libraries, like https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/training_loop.py#L341.

Can you open an issue there so they update their code? @alsrgv @williamFalcon

**Comment by vadimkantorov** [comment]
@vincentqb About the collections.Counter: I think it's a good idea to remove the usage of collections.Counter there, also because it is now present in `state_dict()` output and if someone implements a custom serialization (e.g. to HDF5), de/serializing a Counter instead of a list is extra burden (especially if it's not breaking too many things to keep the state_dict() output simpler).

**Comment by vincentqb** [comment]
> @vincentqb About the collections.Counter: I think it's a good idea to remove the usage of collections.Counter there, also because it is now present in `state_dict()` output and if someone implements a custom serialization (e.g. to HDF5), de/serializing a Counter instead of a list is extra burden (especially if it's not breaking too many things to keep the state_dict() output simpler).

@vadimkantorov -- That's a good point. For serialization, we could use a sorted list of tuples instead. Thoughts? Would you be interested in opening a pull request for this?

For now, we have #33356 as a simple fix for the issue here.

**Comment by vadimkantorov** [comment]
@vincentqb a sortes list of tuples or plain dict is nice! (especially if it’s not a separate structure) + resorting on load_state_dict would work as well? 

Unfortunately I’m very slow in PyTorch development... When I‘ll have the time for this again, I’m working on implementing inplace batchnorm. So unfortunately I’m not much of help for now, only complaints :)
